arch/rv32i/src/syscall.rs

```diff
@@ -130,221 +130,226 @@ impl kernel::syscall::UserspaceKernelBoundary for SysCall {
-          lw   x1,  3*4(sp)
-          lw   x3,  4*4(sp)
-          lw   x4,  5*4(sp)
-          lw   x5,  6*4(sp)
-          lw   x6,  7*4(sp)
-          lw   x7,  8*4(sp)
-          lw   x8,  9*4(sp)
-          lw   x9,  10*4(sp)
-          lw   x10, 11*4(sp)
-          lw   x11, 12*4(sp)
-          lw   x12, 13*4(sp)
-          lw   x13, 14*4(sp)
-          lw   x14, 15*4(sp)
-          lw   x15, 16*4(sp)
-          lw   x16, 17*4(sp)
-          lw   x17, 18*4(sp)
-          lw   x18, 19*4(sp)
-          lw   x19, 20*4(sp)
-          lw   x20, 21*4(sp)
-          lw   x21, 22*4(sp)
-          lw   x22, 23*4(sp)
-          lw   x23, 24*4(sp)
-          lw   x24, 25*4(sp)
-          lw   x25, 26*4(sp)
-          lw   x26, 27*4(sp)
-          lw   x27, 28*4(sp)
-          lw   x28, 29*4(sp)
-          lw   x29, 30*4(sp)
-          lw   x30, 31*4(sp)
-          lw   x31, 32*4(sp)
-
-          addi sp, sp, 34*4   // Reset kernel stack pointer
-          "
-
-          :
-          : "r"(state as *mut Riscv32iStoredState)
-          : "memory"
-          : "volatile");
+        asm!(
+            // Before switching to the app we need to save the kernel registers to
+            // the kernel stack. We then save the stack pointer in the mscratch
+            // CSR (0x340) so we can retrieve it after returning to the kernel
+            // from the app.
+            //
+            // A few values get saved to the kernel stack, including an app
+            // register temporarily after entering the trap handler. Here is a
+            // memory map to make it easier to keep track:
+            //
+            // ```
+            // 34*4(sp):          <- original stack pointer
+            // 33*4(sp):
+            // 32*4(sp): x31
+            // 31*4(sp): x30
+            // 30*4(sp): x29
+            // 29*4(sp): x28
+            // 28*4(sp): x27
+            // 27*4(sp): x26
+            // 26*4(sp): x25
+            // 25*4(sp): x24
+            // 24*4(sp): x23
+            // 23*4(sp): x22
+            // 22*4(sp): x21
+            // 21*4(sp): x20
+            // 20*4(sp): x19
+            // 19*4(sp): x18
+            // 18*4(sp): x17
+            // 17*4(sp): x16
+            // 16*4(sp): x15
+            // 15*4(sp): x14
+            // 14*4(sp): x13
+            // 13*4(sp): x12
+            // 12*4(sp): x11
+            // 11*4(sp): x10
+            // 10*4(sp): x9
+            //  9*4(sp): x8
+            //  8*4(sp): x7
+            //  7*4(sp): x6
+            //  6*4(sp): x5
+            //  5*4(sp): x4
+            //  4*4(sp): x3
+            //  3*4(sp): x1
+            //  2*4(sp): _return_to_kernel (address to resume after trap)
+            //  1*4(sp): *state   (Per-process StoredState struct)
+            //  0*4(sp): app s0   <- new stack pointer
+            // ```
+
+            // Move the stack pointer down to make room.
+            "addi sp, sp, -34*4",
```

Comment:
Why add all of the `"",` on each line?

Reply:
I've seen some examples of other people using `asm!()` doing that. It has the added benefit of my editor and rustfmt aligning the instructions correctly and detecting whether I'm using the correct, spaces-based indentation. I think either would be fine with me though.

Reply:
[The Rust unstable book seems to prefer this syntax over multiline strings](https://doc.rust-lang.org/nightly/unstable-book/library-features/asm.html#inputs-and-outputs), for what it's worth.

Reply:
I do wish they would add more examples to the book. The longest assembly block there is just two instructions.

Reply:
I agree that the documentation is very sparse. Reading through the documentation again seems that using multiple strings was chosen exactly for the benefits I've noticed while using it. To cite them: _"First, we can see that asm! allows multiple template string arguments; each one is treated as a separate line of assembly code, as if they were all joined together with newlines between them. This makes it easy to format assembly code."_

As I said both is fine with me. Should I change it to be a multiline string?

Reply:
I'm typically on the side of smaller commits are better. However, I'm also a believer that if rustfmt doesn't enforce it then it's completely up to the author. I am worried about a collision with my stack pointer PR however :)